### PR TITLE
Weather years script

### PR DESCRIPTION
**Pull request by louispt1**
Primary changes:
- Added weather years script which reads input files to generate weather years profiles and upload them. The script also allows end users to upload custom weather profiles directly for multiple scenarios at once.
- Included building profile heating generator based on pre-existing methods in ETDataset.

### CODE REVIEW

**Comment by kndehaan** [comment]
TODO @louispt1 : check potential fall out of scenario_from_csv.py

**Comment by kndehaan** [comment]
I'm testing to add weather curves that should be uploaded directly (no generation needed) according to how the manual describes it should be done:
image

The curves are however not uploaded to the scenario. Basically, this function is not necessary for the weather years tool since users can upload custom curves with the scenario_from_csv tool. I would suggest to not add this functionality to weather years tools as well, therefore removing this part from the documentation (I can pick this up) and also cleanup the script where this uploading of 'other' weather curves was attempted (todo Louis). Rather than trying to fixing this functionality, Do you agree? @louispt1

**Comment by louispt1** [comment]
> I'm testing to add weather curves that should be uploaded directly (no generation needed) according to how the manual describes it should be done (offshore wind and air temperature): image
> 
> The curves are however not uploaded to the scenario. Basically, this function is not necessary for the weather years tool since users can upload custom curves with the scenario_from_csv tool. I would suggest to not add this functionality to weather years tools as well, therefore removing this part from the documentation (I can pick this up) and also cleanup the script where this uploading of 'other' weather curves was attempted (todo Louis). Rather than trying to fixing this functionality, Do you agree? @louispt1

Makes sense! Yes it should be very quick to remove this from the weather years script

**Comment by kndehaan** [comment]
The weather years functionality doesn't work as expected yet when running the scenario_from_csv tool. 

- It only generates the insulation curves, so not the agriculture and buildings heat demand
- It uploads these generated curves in the input folder (e.g. input/curves/dummy_weather_data), whereas I would like to have it added to output/curves/short_scenario_name)
- The generated curves are not uploaded to the scenario

@louispt1 can you look into this?

**Comment by mabijkerk** [comment]
> With this new functionality, is it possible to still only supply temperature, radiation and thermostat data? Or are wind_speed and G2A parameters mandatory?

Can't we have the users specify which profiles they want to generate?

This might be done based on the input they leave in the input folder when running the script: if they only leave in wind speeds, it only generates wind profiles, if they leave in wind speeds and temperature, radiation and thermostat data then it generates heat profiles and wind profiles.

This would offer the users full flexibility to determine which curves to generate. I think it is similar to how we config the scenario tools for extracting data-exports for example. I am open to other ideas however. Let me know what you think @kndehaan @louispt1.

**Comment by kndehaan** [comment]
> This might be done based on the input they leave in the input folder when running the script: if they only leave in wind speeds, it only generates wind profiles, if they leave in wind speeds and temperature, radiation and thermostat data then it generates heat profiles and wind profiles.

Good suggestion. So build in a check on what input data is provided, if all input data is required for generating insulation_xxx profiles (irradiation, temperature, thermostat), then generate these profiles. If all input data required for generating agriculture / buildings heat demand, then generate these curves. If there's input data missing to generate something at all, than feedback should be printed to the terminal that input data is missing end perhaps refer to docs. 

(Note that wind speeds should be provided for agriculture / buildings heat demand, the tool doesn't generate wind profiles)

Do you think you can implement this @louispt1 ?

**Comment by louispt1** [comment]
@noracato Can you look at [9ca2c50](https://github.com/quintel/scenario-tools/pull/42/commits/9ca2c505c30aa188869d7b7e3691ab0fff628a1e)? I am struggling to find a simple way to get both  scenario.set_heat_demand_curves() and scenario.set_building_agriculture_curves() to work. Either I exclude buildings and agriculture and I get all the insulation curves fine, or it searches for the building and ag curves without the input_curves_folder prepended to the path. This is definitely an issue with how I've set it up, but I'm wondering if there's a simpler solution than refactoring the whole buildings and ag class?

**Review by kndehaan — APPROVED** [review]
Looking good!

File: helpers/Scenario.py [hunk 1]
```diff
@@ -93,29 +99,85 @@ def properties_as_json(self):
...
     def set_heat_demand_curves(self):
-        '''
-        Checks if a heat_demand folder was supplied, and sets self.heat_demand_curves
-        accordingly.
-
-        self.heat_demand is a folder inside the input curves folder
-
-        self.heat_demand_curves links to a generator method, and can thus generate all 15
-        heat demand curves (Curve) iteratively
-        '''
```
Comment (noracato) [inline helpers/Scenario.py:106]: Let's keep the function comments :)

File: helpers/Scenario.py [hunk 2]
```diff
@@ -93,29 +99,85 @@ def properties_as_json(self):
...
+        elif curves_length == 12:
+            # Load temperature, wind speed, and parameters only
+            temp = self._load_heat_data(read_heat_demand_input, input_folder, 'temperature')
+            wind_speed = self._load_heat_data(read_heat_demand_input, input_folder, 'wind_speed')
+            parameters = self._load_heat_data(load_g2a_parameters, input_folder)
+        elif curves_length == 2:
+            # Load temperature, irradiation, and thermostat data
+            temp = self._load_heat_data(read_heat_demand_input, input_folder, 'temperature')
+            irr = self._load_heat_data(read_heat_demand_input, input_folder, 'irradiation')
+            therm = self._load_heat_data(read_thermostat, input_folder)
```
Comment (noracato) [inline helpers/Scenario.py:128]: This does not belong on the Scenario object

File: helpers/Scenario.py [hunk 3]
```diff
@@ -93,29 +99,85 @@ def properties_as_json(self):
...
+                    data = loader_function(self.heat_demand)
+                else:
+                    data = loader_function(self.heat_demand, data_type)
+                return data
+            except FileNotFoundError as e:
+                logging.warning(f"File not found: {e.filename}")
+                return None
+            except IOError as e:
+                logging.warning(f"I/O error({e.errno}): {e.strerror}")
+                return None
```
Comment (noracato) [inline helpers/Scenario.py:150]: This is a weather helper, not something part of a scenario

**Review by noracato — APPROVED** [review]
LGTM! We can do a session one day to talk about how to structure objects in Python. When you get better in defining structures and objects, programming end debugging also becomes a lot easier.

Don't forget to add some info to the docs @louispt1 and @kndehaan

**Pull request merged**